srcts/src/components/webcomponents/layoutColumns.ts

```diff
@@ -0,0 +1,560 @@
+ * | 2 | 2 | . | 4 | 4 | 4 | 4 |   |   |   |   | ...
+ * ```
+ *
+ * Because we recycle column widths to match the number of kids, we can't
+ * guarantee that the pattern repeats by row. To quickly summarize:
+ *
+ * * Each content item has a width (`widths[idx]`) and empty space before
+ *   the item: `before[idx]` + `after[idx - 1]` (the space before this item
+ *   plus the space _after_ the previous item).
+ * * We maintain a cursor that knows the 0-indexed column position. At each
+ *   step we:
+ *   * Move the cursor forward by the empty space before the item
+ *   * Decide if we require a starting class (`g-start-{break}-{cursor + 1}`)
+ *   * Add starting class and content width class (`g-start-{break}-{width}`)
+ *     for the item
+ *   * Move the cursor forward by the width of the item.
+ *
+ * We take into account a few edge cases:
+ *
+ * * We *don't need* a starting class if the item would naturally reflow to the
+ *   next row.
+ * * We *do need* a starting class if the item would fit into the empty space
+ *   of the current row, but there isn't enough room for the item _after_
+ *   accounting for blocked empty space.
+ * * If adding empty space causes a new row, but adding the content item
+ *   would cause _another row break_, we skip the empty row.
+ *
+ * @see https://getbootstrap.com/docs/5.3/layout/breakpoints/
+ *
+ * @param breaks The BreakpointColumnSpec.
+ * @param elements The children of the custom element.
+ * @param unitsRow The number of column units in a row.
+ */
+function writeGridClasses(
+  breaks: BreakpointColumnSpec,
+  elements: HTMLCollectionOf<Element>,
+  unitsRow = 12
+): void {
+  // breaks are column breakpoints
+  const nKids = elements.length;
+
+  function addClassToEl(idx: number, newClass: string) {
+    elements[idx].classList.add(newClass);
+  }
+
+  for (const [breakName, bk] of breaks) {
+    if (bk.width.length > nKids) {
+      const msg = `Truncating number of widths at '${breakName}' breakpoint to match number of elements.`;
+      console.warn(msg, { widths: bk.width.length, elements: nKids });
+    }
+
+    // Find widths, before, after for each element, recycled from the spec
+    const widths = recycleToLength(bk.width, nKids);
+    const before = recycleToLength(bk.before, nKids);
+    const after = recycleToLength(bk.after, nKids);
+
+    let cursor = 0;
+    const updateCursor = (incr: number, isEmpty = false) => {
+      cursor = Math.abs(cursor);
+      let newCursor = cursor + incr;
+      if (newCursor == unitsRow) {
+        // we've reached the final column, allow for a natural break
+        newCursor = 0;
+      }
+      if (newCursor > unitsRow) {
+        // this row is full, empty columns are allowed to break to the next row
+        // which we signal with cursor < 0
+        newCursor = isEmpty ? -1 * (newCursor % unitsRow) : incr;
+      }
+      // console.log(`cursor: ${cursor} -> ${newCursor} (+${incr}${isEmpty ? " (empty)" : ""})`);
```

Comment:
```suggestion
```

Reply:
I intentionally kept this comment. If our future selves need to debug this code, this is the most succinct way to get a sense of how this whole thing works. That felt non-obvious enough that I think it's worth keeping the comment for later.